SEED-platform/seed
# #2477 feat: run bsyncr analysis

#### Any background context you want to provide?
SEED is working on adding ability for users to run analyses on properties. The first analysis available will be bsyncr.

#### What's this PR do?
- Adds pipeline steps for running bsyncr analysis (once it's been prepared)
- Adds endpoint `/analyses/<id>/start` to kickoff this process

#### How should this be manually tested?
To fully test this, you'll need to get bsyncr-server up and running locally.

1. get bsyncr running with `docker run --network="seed_default" --name="bsyncr-server" -e NOAA_TOKEN=<your noaa token> seedplatform/bsyncr-server:latest`. You can get a noaa token here: https://www.ncdc.noaa.gov/cdo-web/token
If you're not using the docker-compose.dev file for running seed, then leave out the `--network=...` argument and replace it with `-p 5000:5000`.
Test that things are working by execing into the web container and running `curl -X POST bsyncr-server:5000`, or `curl -X POST localhost:5000`. You should get a response telling you to include a file.

2. Configure your SEED instance. If using docker-compose, edit your .env file to include the following
```bash
BSYNCR_SERVER_HOST='bsyncr-server'
BSYNCR_SERVER_PORT='5000'
```
If you're running SEED without docker-compose, you need to export these two environment variables (replacing 'bsyncr-server' with 'localhost') yourself and restart your celery workers and server (the celery workers part is probably unnecessary if you're doing the next step)
Then configure celery in your local_untracked.py with this to make things simpler
```python
CELERY_BROKER_BACKEND = 'memory'
CELERY_TASK_ALWAYS_EAGER = True
CELERY_TASK_EAGER_PROPAGATES = True
```

3. Run a script to setup and start the analysis. Download this script here: https://gist.github.com/macintoshpie/6a74540ed37bb6df8ecf847109170e5e
Run it with `python manage.py shell < test_bsyncr.py` (obviously replace the filename path with whatever you name it). Of course you'll need to exec into the web container if using docker-compose. This script will create a new user, org, property views, meters, and analysis before running the analysis (so maybe do it on a db you don't care about). If everything ran correctly, you should be able to check that the analysis property views were updated and the output files were created
```python
a = Analysis.objects.all().last()
apv = AnalysisPropertyView.objects.filter(analysis=a)
print(apv[0].parsed_results)
```

#### What are the relevant tickets?
#2477 

#### Screenshots (if appropriate)
